### PR TITLE
Portable Span IndexOf for OrdinalIgnoreCase

### PR DESCRIPTION
**Pull request by AlexRadch**
I added Portable Span IndexOf for OrdinalIgnoreCase. Close https://github.com/dotnet/corefx/issues/28336 issue. 

I did not found any performance tests for any Span methods with OrdinalIgnoreCase comparison.

Should I add such test inly for Span IndexOf or for other Span methods also?

### CODE REVIEW

**Comment by ahsonkhan** [comment]
>Should I add such test inly for Span IndexOf 

Yes please and share the performance improvement numbers. Compare before/after for IndexOf (ordinal ignore case) for a variety of length spans: 1, 10, 100, 1000.

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs [hunk 1]
```diff
@@ -139,10 +139,138 @@ public static int IndexOf(this ReadOnlySpan<char> span, ReadOnlySpan<char> value
...
+                            {
+                                a0++;
+                                lengthA--;
+                                continue;
+                            }
+
+                            // First char found evaluate CompareTo with OrdinalIgnoreCase
+
+                            // Pointers to second chars and length of remaining value
+                            char* a = a0 + 1;
```
[the remark this answers is not in the document]
Comment (AlexRadch) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: I does not understand your question? How can I doesn't use pointer here?

**Comment by danmosemsft** [comment]
> Compare before/after for IndexOf (ordinal ignore case) for a variety of length spans: 1, 10, 100, 1000.

For both the case of the hit being found at the start, and at the end or not at all.

**Comment by ahsonkhan** [comment]
Yes please. In addition, data to measure the overhead of the optimization in the worst case would be great. Measure the scenario, I called out in the original PR: https://github.com/dotnet/corefx/pull/28239#issue-176098021

**Comment by ahsonkhan** [comment]
> For both the case of the hit being found at the start, and at the end or not at all

In the past, I have used length/2 as the "found" location as that covers the ground for finding at start (when length is 1) and effectively "near the end" for large lengths like 1000, etc. This way we have fewer tests but still get the sense of the perf characteristics. Maybe that approach can be applied here.

**Comment by AlexRadch** [comment]
I added performance test but I can not run it!

I use next command `msbuild /t:BuildAndTest /p:Performance=true /p:ConfigurationGroup=Release /p:TargetOS=Windows_NT /p:TargetGroup=netfx` but it does nothing.

**Comment by AlexRadch** [comment]
Next command `msbuild /t:BuildAndTest /p:Performance=true /p:ConfigurationGroup=Release /p:TargetOS=Windows_NT` runs performance tests but I should run them for portable span for netfx TargetGroup.

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs [hunk 2]
```diff
@@ -88,6 +88,53 @@ public void StringIndexOfChar(int size)
...
+            }
+
+            int index = 0;
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                    {
+                        index |= str.IndexOf("5", comparisonType);
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: This is measuring string.IndexOf, not span.IndexOf.

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs [hunk 3]
```diff
@@ -88,6 +88,53 @@ public void StringIndexOfChar(int size)
...
+        [InlineData(100, StringComparison.OrdinalIgnoreCase)]
+        [InlineData(1000, StringComparison.CurrentCulture)]
+        [InlineData(1000, StringComparison.CurrentCultureIgnoreCase)]
+        [InlineData(1000, StringComparison.InvariantCulture)]
+        [InlineData(1000, StringComparison.InvariantCultureIgnoreCase)]
+        [InlineData(1000, StringComparison.Ordinal)]
+        [InlineData(1000, StringComparison.OrdinalIgnoreCase)]
+        public void IndexOf_StringComparison(int size, StringComparison comparisonType)
+        {
+            string str = new string('0', size / 2) + "5";
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: This is a good baseline test, but we should also consider measuring scenarios with culture specific characters (not only ascii) and upper and lower case characters together.

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs (hunk 2 above)
Comment (AlexRadch) [reply src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: Oh sorry

File: src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs (hunk 3 above)
Comment (AlexRadch) [reply src/System.Memory/tests/Performance/Perf.Span.IndexOf.cs]: I can not run any performance test for netfx ;(

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (morganbr) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: You could slice a's span instead and use normal indexing with b's span. That would also fix the current out of bounds bug this has when both strings are 0 length. Something like:

```
int aPos = 0;
while (lengthA >= 0)
{
    char charA = strA[aPos];
    if( charA != charB0)
    {
        aPos++;
        lengthA--;
        continue;
     }

int curChar = 0;
ReadOnlySpan<char> aSpan = strA.Slice(start: aPos, length: lengthB);
while (length != 0)
{
    char charA = aSpan[curChar];
    char charB  = strB[curChar];
    if (charA != charB)
    {
         ....
     }
     curChar++;
     length--;
}
```

[the remark this answers is not in the document]
Comment (AlexRadch) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: I just copied CompareToOrdinalIgnoreCase method above. And I didn't run performance tests yet for `netfx`. I got instructions how to run `netfx` performance tests recently and it is look hard.

I will try both ways after I understand how to run performance test for `netfx`.

**Comment by stephentoub** [comment]
@ahsonkhan, what's the status of this PR?  Should it be merged? Closed?  Remaining feedback to be updated?

**Comment by ahsonkhan** [comment]
> @ahsonkhan, what's the status of this PR?  Should it be merged? Closed?  Remaining feedback to be updated?

I wouldn't merge it until we have performance numbers that show the issue has been resolved by this change. I know that @AlexRadch was blocked on running performance tests for netfx (to check portable span) and I suggested a workaround here: https://github.com/dotnet/corefx/issues/29277#issuecomment-384071332.

@AlexRadch, if you are having difficulty with that, please let me know and I can try to help.

Alternatively, we could merge the performance tests only, for now.

**Comment by karelz** [comment]
@AlexRadch any update on the perf measurements?

**Comment by stephentoub** [comment]
The portable implementation was removed, so I believe this can be closed. @viktorhofer, @ahsonkhan?

**Comment by ViktorHofer** [comment]
Right. @AlexRadch thanks for your contribution. Meanwhile we removed the portable implementation as System.Memory will be inbox in the different .NET Standard vNext implementations. We don't want to update the existing package if not absolutely necessary like for bugfixes. Thanks!

**Pull request closed without merging**